src/Qowaiv.DomainModel/EventBuffer.cs

```diff
@@ -99,45 +102,27 @@ public EventBuffer(TId aggregateId, int version)
         /// The event to add.
         /// </param>
         public EventBuffer<TId> Add(object @event)
-        {
-            Guard.NotNull(@event, nameof(@event));
-            buffer.Add(@event);
-            return this;
-        }
+            => new(AggregateId, offset, CommittedVersion, buffer.Add<object>(@event));
 
         /// <summary>Adds  events to the event buffer.</summary>
         /// <param name="events">
         /// The events to add.
         /// </param>
+        [Obsolete("Use Add(event) instead.")]
         public EventBuffer<TId> AddRange(IEnumerable<object> events)
-        {
-            Guard.NotNull(events, nameof(events));
-
-            var index = 0;
-
-            foreach (var @event in events)
-            {
-                Guard.NotNull(@event, $"events[{index++}]");
-                buffer.Add(@event);
-            }
-            return this;
-        }
+            => new(AggregateId, offset, CommittedVersion, buffer.Add(events));
 
         /// <summary>Marks all events as being committed.</summary>
         public EventBuffer<TId> MarkAllAsCommitted()
-        {
-            CommittedVersion = Version;
-            return this;
-        }
+            => new(AggregateId, offset, Version, buffer);
 
         /// <summary>Removes the committed events from the buffer.</summary>
         public EventBuffer<TId> ClearCommitted()
-        {
-            var delta = CommittedVersion - offset;
-            buffer.RemoveRange(0, delta);
-            offset += delta;
-            return this;
-        }
+            => new(
+                aggregateId: AggregateId,
+                offset: CommittedVersion,
```

Comment:
is offset always the same as committedVersion?

Reply:
For ClearCommitted that is the case.